### PR TITLE
Datasources - Part 1: Rework datasource classes

### PR DESCRIPTION
**Pull request by rolnico**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
Feature

**What is the current behavior?**
All the existing `DataSources` based on files extend `FileDataSources`. The difference between datasources considering files in a directory and the datasources  considering files in an archive is not clear enough.

**What is the new behavior (if this is a feature change)?**
Two types of `DataSources` based on files now exists:

- `DirectoryDataSource` (inherited by `GzDirectoryDataSource`, `Bzip2DirectoryDataSource`, `ZstdDirectoryDataSource`, `XZDirectoryDataSource`): considers files in a directory
- `AbstractArchiveDataSource` (only `ZipArchiveDataSource` for now but a `TarArchiveDataSource` could be added later): considers files in an archive.

**Does this PR introduce a breaking change or deprecate an API?**
- [X] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**
- [X] The *Breaking Change* or *Deprecated* label has been added
- [X] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**
Two types of `DataSources` based on files now exists:

- `DirectoryDataSource` (inherited by `GzDirectoryDataSource`, `Bzip2DirectoryDataSource`, `ZstdDirectoryDataSource`, `XZDirectoryDataSource`): considers files in a directory
- `AbstractArchiveDataSource` (only `ZipArchiveDataSource` for now but a `TarArchiveDataSource` could be added later): considers files in an archive.

Therefore, multiple classes have changed, please update your code if you used them directly:
| Old classes | New classes     |
|----------|------------- |
| `Bzip2FileDataSource` | `Bzip2DirectoryDataSource`     |
| `FileDataSource` | `DirectoryDataSource`     |
| `GzFileDataSource` | `GzDirectoryDataSource`     |
| `XZFileDataSource` | `XZDirectoryDataSource`     |
| `ZipFileDataSource` | `ZipArchiveDataSource`     |
| `ZstdFileDataSource` | `ZstdDirectoryDataSource`     |

*Note:* Their constructors are not the same so you will have to adapt your code to the new ones, depending on which DataSource you use.

**Other information**:

### CODE REVIEW

**Comment by rolnico** [comment]
This is the first part of the division of the PR https://github.com/powsybl/powsybl-core/pull/2959

File: commons/src/main/java/com/powsybl/commons/datasource/DirectoryDataSource.java [hunk 1]
```diff
@@ -20,40 +20,39 @@
...
-        this.observer = observer;
+    public DirectoryDataSource(Path directory, String baseName,
+                               DataSourceObserver observer) {
+        this(directory, baseName, null, observer);
     }
 
-    @Override
-    public String getBaseName() {
-        return baseName;
+    DirectoryDataSource(Path directory, String baseName,
```
Comment (jonenst) [inline commons/src/main/java/com/powsybl/commons/datasource/DirectoryDataSource.java:36]: add also
```java
DirectoryDataSource(Path directory, String baseName, CompressionFormat compressionFormat) 
```
?

File: commons/src/main/java/com/powsybl/commons/datasource/AbstractArchiveDataSource.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Check if a file exists in the archive. The file name will be constructed as:
+     * {@code <basename><suffix>.<ext>}</p>
+     * @param suffix Suffix to add to the basename of the datasource
+     * @param ext Extension of the file (for example: .iidm, .xml, .txt, etc.)
+     * @return true if the file exists, else false
+     */
+    @Override
+    public boolean exists(String suffix, String ext) throws IOException {
```
[the remark this answers is not in the document]
Comment (rolnico) [reply commons/src/main/java/com/powsybl/commons/datasource/AbstractArchiveDataSource.java]: moved

File: commons/src/test/java/com/powsybl/commons/datasource/AbstractArchiveDataSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        // File
+        Path path = testDir.resolve(archiveWithSubfolders);
+        Files.createFile(path);
+
+        // Create the datasource
+        DataSource dataSource = DataSource.fromPath(path);
+
+        UnsupportedOperationException exception = assertThrows(UnsupportedOperationException.class, () -> {
+            try (OutputStream ignored = dataSource.newOutputStream("foo.bar", true)) {
+                fail();
```
Comment (jonenst) [inline commons/src/test/java/com/powsybl/commons/datasource/AbstractArchiveDataSourceTest.java:78]: no need to fail explicitly ?

File: commons/src/test/java/com/powsybl/commons/datasource/ZipArchiveDataSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    @Override
+    protected String getFileName(String baseName, String mainExtension, ArchiveFormat archiveFormat,
+                                 CompressionFormat compressionFormat) {
+        return testDir + "/" + baseName + (mainExtension.isEmpty() ? "" : ".") + mainExtension
+            + (compressionFormat == null ? "" : "." + compressionFormat.getExtension());
+    }
+
+    static Stream<Arguments> provideArgumentsForWriteThenReadTest() {
+        return Stream.of(
+            Arguments.of("foo", "iidm", CompressionFormat.ZIP),
```
Comment (jonenst) [inline commons/src/test/java/com/powsybl/commons/datasource/ZipArchiveDataSourceTest.java:96]: should use the same semantics for the "mainExtension" as other test classes

            Arguments.of("foo", "iidm", CompressionFormat.ZIP),
vs
            Arguments.of("foo", ".iidm", CompressionFormat.ZIP),

Comment (jonenst) [reply commons/src/test/java/com/powsybl/commons/datasource/ZipArchiveDataSourceTest.java:96]: (or just remove the mainExtension concept from this PR entirely, it's just used for tests and you can put everything in one string)

File: commons/src/test/java/com/powsybl/commons/datasource/AbstractArchiveDataSourceTest.java (hunk 3 above)
Comment (rolnico) [reply commons/src/test/java/com/powsybl/commons/datasource/AbstractArchiveDataSourceTest.java:78]: It's not needed explicitly but it need something in the try() that cannot throw a runtime exception

File: commons/src/test/java/com/powsybl/commons/datasource/ZipArchiveDataSourceTest.java (hunk 4 above)
Comment (rolnico) [reply commons/src/test/java/com/powsybl/commons/datasource/ZipArchiveDataSourceTest.java]: done

**Review by jonenst — APPROVED** [review]

**Review by flo-dup — APPROVED** [review]

**Pull request merged**